File: webui/src/main/scala/com/taskadapter/reporting/StacktraceCleaner.scala
package com.taskadapter.reporting

/**
  * strips some irrelevant stacktrace elements, like Java internals
  */
object StacktraceCleaner {

  private val strings = Set("java.util.concurrent", "java.lang.Thread")
  private val regexp = strings.map(s => s".*($s).*").mkString("|")

  val replacement = "/./"

  def stripInternalStacktraceItems(stackTrace: String) : String = {
    stackTrace
      .split(System.lineSeparator())
      .map(line =>  {
        val trimmed = line.trim

        if (trimmed.matches(regexp)) {
          replacement
        } else {
          line
        }
      })
      .mkString(System.lineSeparator())
  }

}

File: webui/src/main/java/com/taskadapter/reporting/StacktraceCleaner.java
package com.taskadapter.reporting;

import java.util.Arrays;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * strips some irrelevant stacktrace elements, like Java internals
 */
public class StacktraceCleaner {
    private static final Set<String> strings = Set.of("java.util.concurrent", "java.lang.Thread");
    private static final String regexp = strings.stream()
            .map(s -> ".*(" + s + ").*")
            .collect(Collectors.joining("|"));

    private static final String replacement = "/./";

    public static String stripInternalStacktraceItems(String stackTrace) {
        var strings = stackTrace
                .split(System.lineSeparator());
        return Arrays.stream(strings)
                .map(line -> {
                    var trimmed = line.trim();

                    if (trimmed.matches(regexp)) {
                        return replacement;
                    } else {
                        return line;
                    }
                })
                .collect(Collectors.joining(System.lineSeparator()));
    }
}
